pd/src/components/shielded_pool.rs

```diff
@@ -408,48 +407,57 @@ pub trait View: WriteOverlayExt {
         Ok(())
     }
 
-    async fn denom_by_asset(&self, asset_id: &asset::Id) -> Result<Option<Denom>> {
+    async fn known_assets(&self) -> Result<KnownAssets> {
         Ok(self
-            .get_proto(format!("shielded_pool/assets/{}/denom", asset_id).into())
+            .get_domain("shielded_pool/known_assets".into())
             .await?
-            .map(|denom: String| {
-                asset::REGISTRY
-                    .parse_denom(&denom)
-                    .expect("tree only records valid denoms")
-            }))
+            .unwrap_or_default())
+    }
+
+    async fn denom_by_asset(&self, asset_id: &asset::Id) -> Result<Option<Denom>> {
+        self.get_domain(format!("shielded_pool/assets/{}/denom", asset_id).into())
+            .await
     }
 
     #[instrument(skip(self))]
-    async fn register_denom(&self, denom: &Denom) {
-        let asset_id = denom.id();
-        tracing::debug!(?asset_id);
-        self.put_proto(
-            format!("shielded_pool/assets/{}/denom", asset_id).into(),
-            denom.to_string(),
-        )
-        .await
+    async fn register_denom(&self, denom: &Denom) -> Result<()> {
+        let id = denom.id();
+        if self.denom_by_asset(&id).await?.is_some() {
+            tracing::debug!(?denom, ?id, "skipping existing denom");
+            Ok(())
+        } else {
+            tracing::debug!(?denom, ?id, "registering new denom");
+            // We want to be able to query for the denom by asset ID...
+            self.put_domain(
+                format!("shielded_pool/assets/{}/denom", id).into(),
+                denom.clone(),
+            )
+            .await;
+            // ... and we want to record it in the list of known asset IDs
+            // (this requires reading the whole list, which is sad, but hopefully
+            // we don't do this often).
+            let mut known_assets = self.known_assets().await?;
+            known_assets.0.push(Asset {
+                id,
+                denom: denom.clone(),
+            });
+            self.put_domain("shielded_pool/known_assets".into(), known_assets)
+                .await;
```

Comment:
without writing a list of all of the known asset IDs, we can't iterate over the denoms, and if we're writing a list anyways, we might as well bundle in the denoms.